### ISSUE TITLE
dagster-dbt does not support `unit_test` nodes and @dbt_asset selection by tag in the same dbt project

### ISSUE THREAD

**Issue by mmutso-boku** [dagster-io/dagster#23821]
### Dagster version

1.7.13

### What's the issue?

I have a dagster + dbt setup where in the dbt manifest I have different nodes - models, seeds, tests and as of recently, unit_tests.
I also have a `@dbt_asset` definition where the `select` parameter is by tag:
```
@dbt_assets(manifest=MY_DBT_PROJECT_MANIFEST_PATH,
            select='tag:my_special_tag')
```

When having a unit_test defined in the dbt project, as specified in https://docs.getdbt.com/reference/resource-properties/unit-tests#before-you-begin, the `@dbt_assets` definition will fail when attempting to load these assets into code location Definitions:
```
../__init__.py:54: in <module>
    assets=[*load_assets_from_package_module(assets)],
../../.venv/lib/python3.10/site-packages/dagster/_core/definitions/load_assets_from_modules.py:270: in load_assets_from_package_module
    ) = assets_from_package_module(package_module)
../../.venv/lib/python3.10/site-packages/dagster/_core/definitions/load_assets_from_modules.py:218: in assets_from_package_module
    return assets_from_modules(
../../.venv/lib/python3.10/site-packages/dagster/_core/definitions/load_assets_from_modules.py:56: in assets_from_modules
    for module in modules:
../../.venv/lib/python3.10/site-packages/dagster/_core/definitions/load_assets_from_modules.py:335: in find_modules_in_package
    submodule = import_module(modname)
/opt/homebrew/Cellar/python@3.10/3.10.14/Frameworks/Python.framework/Versions/3.10/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
<frozen importlib._bootstrap>:1050: in _gcd_import
    ???
<frozen importlib._bootstrap>:1027: in _find_and_load
    ???
<frozen importlib._bootstrap>:1006: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:688: in _load_unlocked
    ???
<frozen importlib._bootstrap_external>:883: in exec_module
    ???
<frozen importlib._bootstrap>:241: in _call_with_frames_removed
    ???
../assets/seeds.py:7: in <module>
    @dbt_assets(manifest=MY_DBT_PROJECT_MANIFEST_PATH,
../../.venv/lib/python3.10/site-packages/dagster/_core/decorator_utils.py:223: in wrapped_with_context_manager_fn
    return fn(*args, **kwargs)
../../.venv/lib/python3.10/site-packages/dagster_dbt/asset_decorator.py:311: in dbt_assets
    ) = build_dbt_multi_asset_args(
../../.venv/lib/python3.10/site-packages/dagster_dbt/asset_utils.py:806: in build_dbt_multi_asset_args
    unique_ids = select_unique_ids_from_manifest(
../../.venv/lib/python3.10/site-packages/dagster_dbt/utils.py:309: in select_unique_ids_from_manifest
    selected, _ = selector.select_nodes(parsed_spec)
../../.venv/lib/python3.10/site-packages/dbt/graph/selector.py:160: in select_nodes
    direct_nodes, indirect_nodes = self.select_nodes_recursively(spec)
../../.venv/lib/python3.10/site-packages/dbt/graph/selector.py:132: in select_nodes_recursively
    bundles = [self.select_nodes_recursively(component) for component in spec]
../../.venv/lib/python3.10/site-packages/dbt/graph/selector.py:132: in <listcomp>
    bundles = [self.select_nodes_recursively(component) for component in spec]
../../.venv/lib/python3.10/site-packages/dbt/graph/selector.py:132: in select_nodes_recursively
    bundles = [self.select_nodes_recursively(component) for component in spec]
../../.venv/lib/python3.10/site-packages/dbt/graph/selector.py:132: in <listcomp>
    bundles = [self.select_nodes_recursively(component) for component in spec]
../../.venv/lib/python3.10/site-packages/dbt/graph/selector.py:130: in select_nodes_recursively
    direct_nodes, indirect_nodes = self.get_nodes_from_criteria(spec)
../../.venv/lib/python3.10/site-packages/dbt/graph/selector.py:84: in get_nodes_from_criteria
    collected = self.select_included(nodes, spec)
../../.venv/lib/python3.10/site-packages/dbt/graph/selector.py:70: in select_included
    return set(method.search(included_nodes, spec.value))
../../.venv/lib/python3.10/site-packages/dbt/graph/selector_methods.py:270: in search
    if hasattr(node, "tags") and any(fnmatch(tag, selector) for tag in node.tags):
E   TypeError: 'NoneType' object is not iterable
```

I did some digging, and found this issue in dbt github https://github.com/dbt-labs/dbt-core/issues/10519 . This is the exact same issue I have, however it also seems to me this originates from dagster, and not dbt.

Going through the stacktrace with a debugger, I found that the `manifest` object in `dagster_dbt.utils.select_unique_ids_from_manifest` is build using some `_DictShim` construct, which seems to take the place of `dbt.contracts.graph.nodes.UnitTestDefinition` in `dbt.graph.selector_methods.SelectorMethod.unit_tests`, which is why the `hasattr(node, "tags")` call works with dbt in `dbt.graph.selector_methods.TagSelectorMethod.search` but not with dagster-dbt, because `UnitTestDefinition` points the `tags` attribute to `config.tags`, but the `_DictShim` does not do that.

At least that is my understanding so far.

### What did you expect to happen?

I expect `@dbt_assets` to work even when I have `"resource_type": "unit_test"` nodes present in the dbt manifest

### How to reproduce?

I don't have an MVP example to provide, but basically have some dbt unit_tests defined in a separate yaml, as shown here https://docs.getdbt.com/reference/resource-properties/unit-tests#before-you-begin

If you were to move this `unit_test` block into the model config directly, then this bug won't surface. As in:
```
version: 2

models:
  - name: <model name>
    unit_tests: ...
```

EDIT: Now thinking about it, this probably works because dbt unit_tests would also stop working as they are no longer recognized when defined this way, hence why no node for them in the manifest this way, but I haven't tested it yet.

### Deployment type

Dagster Helm chart

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by maximearmstrong** [comment]
Hi @mmutso-boku - thanks for reaching out!

This was fixed in #23232 and included in Dagster [1.8.0](https://github.com/dagster-io/dagster/releases/tag/1.8.0).

**Comment by lamalex** [comment]
Working for me in 1.8 thank you!
